kolena/_experimental/dataset/_dataset.py

```diff
@@ -83,6 +86,40 @@ def _infer_datatype_value(x: str) -> str:
     return DatapointType.TABULAR.value
 
 
+def _add_datatype(df: pd.DataFrame, sep: str = ".") -> None:
+    """Adds `data_type` column(s) to input DataFrame."""
+    prefixes = {
+        column.rsplit(sep=sep, maxsplit=1)[0]
+        for column in df.columns.values
+        if isinstance(column, str) and sep in column
+    }
+    if prefixes:
+        df[DATA_TYPE_FIELD] = DatapointType.COMPOSITE.value
+        _format_composite(df, prefixes, sep)
+    else:
+        df[DATA_TYPE_FIELD] = _infer_datatype(df)
+
+
+def _format_composite(df: pd.DataFrame, prefixes: Set[str], sep: str) -> None:
```

Comment:
Not really a fan of the arguments here, but `_add_datatype` method was getting bulky. I'll likely update this to work on a single composite (prefix at a time and keep the for loop in `_add_datatype`